### DISCUSSION TITLE
Map low resolution colour image information into (same copy) high resolution black and white image to deliver hi res colour image

### DISCUSSION THREAD

**Discussion by rleaver152** [ImageMagick/ImageMagick#2299] [Help] [answered]
Apologies if a FAQ.   I have a black and white TIF image high resolution and a colour png low resolution of the same image.   I would like to map the colour information as optimally as possible into the higher res black and white image so I end up with the best possible high resolution colour image,  and wondered if there is a smart way of using ImageMagick to do this?    

All image conversions will be lossless so not so worried about formats.

Thanks in advance.

**Comment by snibgo** [comment]
I suppose the "black and white" version is grayscale?

Perhaps: resize the colour version to the size of the grayscale. Then use the colours for the H and C channels of HCL, or a and b channels of Lab.

**Reply by rleaver152** [reply]
Many thanks - yes grayscale,   I'm not familiar with the best way to do the HC channel mapping - grateful for any tips

**Comment by snibgo** [comment]
What version of IM, on what platform?

Please supply sample input images.

**Reply by rleaver152** [reply]
Mac Mojave - bash script.   samples to follow

Version: ImageMagick 7.0.10-19 Q16 x86_64 2020-06-12 https://imagemagick.org
Mac Mini late 2014 with external AMD GPU RX580

**Comment by fmw42** [comment, marked as answer]
Can you post your images or if too big, small subsections or other samples?

**Reply by rleaver152** [reply]
Here they are:

https://we.tl/t-X2R5W4OFkw
(5 day link with originals)

![img3657_original_small_colour](https://user-images.githubusercontent.com/19638129/87969504-2d3e3200-caba-11ea-932f-d0a6d28740b3.jpg)
![img3657_original_small](https://user-images.githubusercontent.com/19638129/87969509-2f07f580-caba-11ea-8e5f-b16423f718c4.png)

![img3657_large](https://user-images.githubusercontent.com/19638129/87969741-9de54e80-caba-11ea-95d2-2486e2000a49.png)


I want to apply the colour from the small file to its larger counterpart.  I also attach the original source b/w small version before it was coloured.  (Before anyone asks,  I am stuck with the small AI coloured sizes and if I resize those,  I get compounded file degradation from that process irrespective of formats input - whereas my b/w resized originals can be pristine in comparison).   The large b/w here is not as large as the ones I am using in practice, but gets under the forum size threshold for posting as an example.

I should also mention that the large b/w pristine enlargements are AI enlarged and so it is not a pixel multiplication process but AI Generative Adversarial Network (GAN) "hallucination of detail"

I really appreciate all help.  thanks again

**Reply by fmw42** [reply]
Here is the way by swapping intensity-like channels in the HCL color space. We resize the color image to the size of the hi res gray image. The convert the color image to HCL color space. Then swap the gray image for the L channel in the HCL and then delete the old L channel from the color image. Then we recombine the H, C and Gray images as HCL. Then convert back to sRGB colorspace.

```
magick hi_res_gray.png -set option:dims "%wx%h" \
\( color.jpg -resize "%[dims]" -colorspace HCL -separate \) \
-swap 0,3 -delete 0 \
-set colorspace HCL -combine -colorspace sRGB \
hi_res_color.png
```

![hi_res_color](https://user-images.githubusercontent.com/13458218/87973555-92bffd80-ca7d-11ea-8fa7-b3fde9240a47.png)


You can do the same in LAB colorspace with pretty much the same result:

```
magick hi_res_gray.png -set option:dims "%wx%h" \
\( color.jpg -resize "%[dims]" -colorspace LAB -separate \) \
-swap 0,1 -delete 0 \
-set colorspace LAB -combine -colorspace sRGB \
hi_res_color_lab.png
```

![hi_res_color_lab](https://user-images.githubusercontent.com/13458218/87973814-f0ece080-ca7d-11ea-802b-f99370ce433e.png)

**Reply by rleaver152** [reply]
Fred,  thank you so much, that is absolutely wonderful work.   Greatly appreciated.  Richard
